
###  Multivariate Analysis
- Learn about Univariate and Multivariate Data Analysis.
- Learn about Encoding and Normalization-Scaling.

&nbsp;

1. Import necessary libraries and Load Data
2. Univariate Data Analysis
3. Checking distribution of data - distplot, violin plot, histograms, commulative distributions
4. Multivariate Data Analysis
4. Encoding Categorical Variables - One-hot encoding and label encoding 
5. Normalization and Scaling 


In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
from PIL import Image 

In [3]:
data = pd.read_csv('Melbourne_housing_FULL.csv')
data.head(3)

Suburb         Address  Rooms Type Method        SellerG      Date  \
0    Abbotsford   68 Studley St      2    h     SS         Jellis  3/9/2016   
1  Airport West   154 Halsey Rd      3    t     PI         Nelson  3/9/2016   
2   Albert Park  105 Kerferd Rd      2    h      S  hockingstuart  3/9/2016   

   Distance  Postcode  Bedroom  ...  Landsize  BuildingArea  YearBuilt  \
0       2.5    3067.0      2.0  ...     126.0           inf        NaN   
1      13.5    3042.0      3.0  ...     303.0           225     2016.0   
2       3.3    3206.0      2.0  ...     120.0            82     1900.0   

                  CouncilArea  Latitude Longtitude             Regionname  \
0          Yarra City Council  -37.8014   144.9958  Northern Metropolitan   
1  Moonee Valley City Council  -37.7180   144.8780   Western Metropolitan   
2   Port Phillip City Council  -37.8459   144.9555  Southern Metropolitan   

   Propertycount      ParkingArea      Price  
0         4019.0          Carport        NaN  
1         3464.0  Detached Garage   840000.0  
2         3280.0  Attached Garage  1275000.0  

[3 rows x 22 columns]

#### This dataset contains 22 columns. For the sake of univariate analysis let us use just one column of data - 'Distance'.

In [4]:
data = data['Distance']

In [5]:
len(data)

34857

#### The first step is to check if the data has any missing values. Some plots will not work in the presence of missing values. So we need to take care of them.

In [6]:
data.isnull().sum()

1

#### There is one missing value out of a total of 34857 observations. For now we can drop that one missing value. Dealing with missing values will be dealt with in a later section.

In [7]:
# drops all missing values in the data
data = data.dropna() 

In [8]:
data.isnull().sum()

0

#  Univariate Analysis


- Univariate analysis refer to the analysis of a single variable. 
- The main purpose of univariate analysis is to summarize and find patterns in the data. 
- The key point is that there is only one variable involved in the analysis.
    
 > Let us take the 'Melbourne Housing' dataset and work on that for the univariate analysis.

####   <span style="font-family: Arial; font-weight:bold;font-size:1.9em;color:#0e92ea">  Now let us check the distribution of the data.
    
#### <span style="font-family: Arial; font-weight:bold;font-size:1.9em;color:#0e92ea"> What does the distribution of data tell us?


<font color=darkblue>
    
* The plot displayed below is called as a Histogram
    
* The X-axis groups the observations from minimum to maximum along the axis on the basis of the discrete points or class intervals

* The Y-axis measures the frequency of occurance of observations for each discrete point or class interval.

In [ ]:
# The following code plots a histrogram using the matplotlib package.
# The bins argument creates class intervals. In this case we are creating 50 such intervals
plt.hist(data, bins=50)

**In the above histogram, the first array is the frequency in each class and the second array contains the edges of the class intervals. These arrays can be assigned to a variable and used for further analysis.**

In [ ]:
sns.distplot(data) # plots a frequency polygon superimposed on a histogram using the seaborn package.
# seaborn automatically creates class intervals. The number of bins can also be manually set.

In [ ]:
sns.distplot(data, hist=False) # adding an argument to plot only frequency polygon

**This distribution can also be visualised in another manner. For this we can use the violin plot function from seaborn. The violin plot shows a vertical mirror image of the distribution along with the original distribution.**

In [ ]:
sns.violinplot(data) # plots a violin plt using the seaborn package.

### Now let us have a closer look at the distribution by plotting a simple histogram with 10 bins.

In [ ]:
plt.figure(figsize=(20,10)) # makes the plot wider
plt.hist(data, color='g') # plots a simple histogram
plt.axvline(data.mean(), color='m', linewidth=1)
plt.axvline(data.median(), color='b', linestyle='dashed', linewidth=1)
plt.axvline(data.mode()[0], color='w', linestyle='dashed', linewidth=1)

<font color=darkblue>
    
- In the above histogram we can see that the second bin with class limits (4.81,  9.62) has the most number of observations. 
- A bulk of the observations lie within the first four classes. The rest of the six classes contain only a very few observations.
- In this case we are measuring distances. We can safely say based on visual observation that most of the distance present on the data are less than 20 kms.
- There are a very few distances that are more than 20 kms. Once we pass the 30 km point the number of observations dlrops further.
- From the above figure we can see the the mean is represented by the magenta line and the mode by the white line . The median is represented by the blue line.
- We can see from the above figure that the mode and the mean are very close to each other and that the median is lower than both.

#### Now we have an idea of how the data is distributed.

# <span style="font-family: Arial; font-weight:bold;font-size:1.9em;color:#0e92ea"> Multivariate analysis

<font color=darkblue>
    
#### Multivariate analysis is performed to understand interactions between different fields in the dataset (or) finding interactions between variables more than 2
#### ex: Pairplot, 3D scatter plot

#### <span style="font-family: Arial; font-weight:bold;font-size:1.9em;color:#0e92ea"> Scatter Plot


<font color=darkblue>

#### A Scatter plot gives us an idea of the association between two variables

In [ ]:
data1 = pd.read_csv('Melbourne_housing_FULL.csv')

In [ ]:
data1.head()

In [ ]:
data1 = data1.dropna()

In [ ]:
sns.pairplot(data1)


**In the above plot scatter diagrams are plotted for all the numerical columns in the dataset. A scatter plot is a visual representation of the degree of correlation between any two columns. The pair plot function in seaborn makes it very easy to generate joint scatter plots for all the columns in the data.**

**A scatter plot can also be plotted for two individual columns:**

In [ ]:
sns.scatterplot(data1['Price'], data1['Distance'])  # Plots the scatter plot using two variables

**The scatter plot only offers visual information about the degree of correlation. In order to obtain more precise information we can use the inbuilt .corr() method in Pandas. This returns a table with all the correlations calculated for the numerical columns.**

In [ ]:
data1.corr()   # displays the correlation between every possible pair of attributes as a dataframe

**The same can be represented graphically using a the heatmap function in seaborn.**

In [ ]:
sns.heatmap(data1.corr(), annot=True)  # plot the correlation coefficients as a heatmap

**Another way of looking at multivariate scatter plot is to use the hue option in the scatterplot() function in seaborn.**

In [ ]:
sns.scatterplot(data1['Distance'], data1['Price'], hue=data1['Type'], palette='Set2') 

**Here the colour of the dots indicates a third variable - Bedroom.**

# <span style="font-family: Arial; font-weight:bold;font-size:1.9em;color:#0e92ea"> Handling non-numeric(Categorical) data

<font color=darkblue>Most of the machine learning models are designed to work on numeric data. Hence, we need to convert categorical text data into numerical data for model building

#### <span style="font-family: Arial; font-weight:bold;font-size:1.9em;color:#0e92ea">One Hot Encoding

<font color=darkblue>One-Hot-Encoding is used to create dummry variables to replace the categories in a categorical variable into features of each category and represent it using 1 or 0 based on the presence or absence of the categorical value in the record

In [ ]:
df_dummies= pd.get_dummies(data1, prefix='Park', columns=['ParkingArea']) #This function does One-Hot-Encoding on categorical text

In [ ]:
df_dummies.head()

In [ ]:
df_dummies.columns   # returns the names of all the columns as a list

#### <span style="font-family: Arial; font-weight:bold;font-size:1.9em;color:#0e92ea"> Sklearn Label Encoding

In [ ]:
from sklearn.preprocessing import LabelEncoder   # import label encoder

labelencoder = LabelEncoder()  

df_dummies['RegionId'] = labelencoder.fit_transform(df_dummies.Regionname) # returns label encoded variable(s)


In [ ]:
df_dummies['RegionId'].head(5)  


In [ ]:
df_dummies['RegionId'].value_counts()   # Number of entries of each class

In [ ]:
df_dummies['RegionId'].unique()   # All the unique classes

In [ ]:
df_dummies['RegionId'].nunique()  # number of unique classes

## Sklearn OneHotEncoder

&nbsp;

<font color=darkblue>
    
   
This function does One-Hot-Encoding on categorical numbers

In [ ]:
from sklearn.preprocessing import OneHotEncoder
hotencoder = OneHotEncoder()
encoded = hotencoder.fit_transform(df_dummies.RegionId.values.reshape(-1,1)).toarray() 
# Returns a numpy array of one hot encoded variables
encoded

In [ ]:
encoded.shape

Let us now add this back to df_dummies dataframe.

In [ ]:
# convert the array into a dataframe. Specifically, one hot encoded dataframe

df_encoded = pd.DataFrame(encoded, columns = ["RegionId_"+str(int(i)) for i in range(encoded.shape[1])])

In [ ]:
df_encoded.head() 

In [ ]:
df_encoded.shape

In [ ]:
df_dummies = pd.concat([df_dummies, df_encoded], axis=1)  # concats two dataframes

In [ ]:
df_dummies.head(5)

<font color=darkblue>
The above steps are just to get comfortable with a few functions. We do not necessarily need to follow those steps in every practice.

# <span style="font-family: Arial; font-weight:bold;font-size:1.9em;color:#0e92ea">Normalization and Scaling


<font color=darkblue>

In this method, we convert variables with different scales of measurements into a single scale. 

<font color=darkblue>
StandardScaler normalizes the data using the formula (x-mean)/standard deviation

In [ ]:
#Scales the data. Essentially returns the z-scores of every attribute

from sklearn.preprocessing import StandardScaler

std_scale = StandardScaler()

std_scale

In [ ]:
data1['Bedroom_Stdscale'] = std_scale.fit_transform(data1[['Bedroom']]) # returns z-scores of the values of the attribute

In [ ]:
data1['Bedroom_Stdscale'].head()

In [ ]:
data1['Bedroom_Stdscale'].min(), data1['Bedroom_Stdscale'].max()  # min and max value

In [ ]:
data1['Bedroom_Stdscale'].describe()   # 5 point summary statistics of the attribute

<font color=darkblue> MinMaxScaler normalizes the data using the formula (x - min)/(max - min)

In [ ]:
from sklearn.preprocessing import MinMaxScaler

minmax_scale = MinMaxScaler()

minmax_scale

In [ ]:
data1['Car_MinMaxScale'] = minmax_scale.fit_transform(data1[['Car']])

In [ ]:
data1['Car_MinMaxScale'].head(5)

In [ ]:
data1['Car_MinMaxScale'].min(), data1['Car_MinMaxScale'].max()

In [ ]:
data1['Car_MinMaxScale'].describe()

<font color=darkblue> Log Transformation 

We are going to write a custom transformer to perform log transformation

In [ ]:
import numpy as np

In [ ]:
# Transform an attribute using a mathematical transformation.
# We might want to do that to change the distribution shape of the data
from sklearn.preprocessing import FunctionTransformer   

log_transformer = FunctionTransformer(np.log1p)

log_transformer

In [ ]:
data1['Distance'].isnull().sum()   # Check number of null values

In [ ]:
data1['Distance']= data1['Distance'].fillna(data1['Distance'].mean())  # impute the null values using the mean

In [ ]:
data1['Distance_logtransform'] = log_transformer.fit_transform(data1[['Distance']])   # Log Transform the attribute 

In [ ]:
data1['Distance_logtransform'].head()

In [ ]:
data1['Distance_logtransform'].describe()    # Descriptive statistics 

**Exponential Transformation** 

In [ ]:
exp_transformer = FunctionTransformer(np.exp) # Exponential transform 

exp_transformer

In [ ]:
data1['Rooms_exptransform'] = exp_transformer.fit_transform(data1[['Rooms']])  #returns the exponential transform of the data

In [ ]:
data1['Rooms_exptransform'].head(5)

In [ ]:
data1['Rooms_exptransform'].describe()